### PR TITLE
Set attestation object containing CBOR encoded authData, fmt and attStmt

### PR DESCRIPTION
**Pull request by yesvivek**
Applications using Javascript WebAuthn APIs to communicate with FIDO2 devices get CBOR encoded attestation object containing authData, fmt and attStmt; more details about the JS API are given at [MDN](https://developer.mozilla.org/en-US/docs/Web/API/AuthenticatorAttestationResponse/attestationObject). Since there are no straight ways to set this value, I have updated libfido to set internal values after parsing the CBOR encoded attestation-object.
This will mean that the applications using libfido doesn't have to handle CBOR data at all, just like how other high level libraries in other languages behave.
Summary of changes:
  - `fido_cred_set_attobj` API is added to set attestation object containing CBOR encoded authData, fmt and attStmt
  - Supporting routines to parse and set the internal structure.
 
Relevant issue #749 .

### CODE REVIEW

File: src/cred.c
Comment (LDVG) [inline src/cred.c:740]: We have a very similar function in `winhello.c`,  see `decode_attobj()`.  If we were to introduce this public API, I feel like it'd make sense to consolidate them.  

One approach would be to move `winhello.c`'s `decode_attobj()` to e.g. `cbor.c` and introduce a `cbor_decode_attobj()` function, like you've done (but I don't think we need it to take a function pointer). Then call this function from both `winhello.c/translate_winhello_cred()` and the new `fido_cred_set_attobj()`.

To ensure a consistent state, `fido_cred_set_attobj()` would have to call `fido_cred_reset_rx()` first.

Any thoughts?

Comment (yesvivek) [reply src/cred.c]: Did not know that winhello had same function, so ended up adding a new one; fixed now.
`fido_cred_reset_rx` was tempting but it clears large-blob too and I don't know what it does. Anyways, now using `fido_cred_reset_rx`.
Pushed new changes according to your suggestion.

Comment (yesvivek) [reply src/cred.c]: Ok, in that case I will add `fido_cred_clean_attobj()`. 
Should `winhello.c/translate_winhello_cred()` also be updated to use it? @LDVG

Comment (LDVG) [reply src/cred.c]: That should not be necessary, `fido_winhello_make_cred()` already calls `fido_cred_reset_rx()` before calling `translate_winhello_cred()`

Comment (yesvivek) [reply src/cred.c]: Ok, pushed a new change to add `fido_cred_clean_attobj()`.

File: src/cred.c
Comment (LDVG) [inline src/cred.c:553]: Would it make sense for `fido_cred_reset_rx()` to call this function?

Comment (yesvivek) [reply src/cred.c:553]: Yes, that makes sense. Pushed it.

**Comment by yesvivek** [comment]
Assuming this change will be merged, any idea when the next release is?

**Review by LDVG — COMMENTED** [review]
I pushed a few additional tweaks on top of your changes, let me know if anything looks off. :slightly_smiling_face: 

> Assuming this change will be merged, any idea when the next release is?

There's no exact dates planned yet. It's likely it'd be sometime this spring.

File: man/fido_cred_set_authdata.3
Comment (yesvivek) [reply man/fido_cred_set_authdata.3]: You are correct, `.Dq` is appropriate here. Pushed the chage.

File: man/fido_cred_set_authdata.3
Comment (yesvivek) [reply man/fido_cred_set_authdata.3]: This is much better, pushed the change.

**Comment by yesvivek** [comment]
@LDVG is there anything needed that is blocking merge with main?

**Comment by LDVG** [comment]
> @LDVG is there anything needed that is blocking merge with main?

Apologies for the delays.  There shouldn't be much more needed to get this through, though a final review may take a little bit longer.  I squashed some of the intermediate commits, I hope that is fine with you.

**Comment by LDVG** [comment]
Note (mostly to myself): Pushed a tentative fuzzer harness; needs seed corpora.

File: fuzz/fuzz_attobj.c
Comment (martelletto) [inline fuzz/fuzz_attobj.c:1]: missing copyright blurb

File: man/fido_cred_set_authdata.3
Comment (martelletto) [inline man/fido_cred_set_authdata.3:164]: it might make sense (at a later point) to add a line or two about the purpose/utility of set_attobj

File: src/cbor.c
Comment (martelletto) [inline src/cbor.c:1154]: something we need to reevaluate at one point is how we deal with duplicate entries in maps, which IIRC are allowed in CTAP CBOR. in some places (like cbor_decode_fmt) we error out, in others (cbor_decode_attstmt, alg case) we silently overwrite previously observed values

**Review by martelletto — COMMENTED** [review]
left some comments; nothing major; LGTM

File: src/cbor.c
Comment (LDVG) [reply src/cbor.c:1154]: Mainly as a reference to myself:

> something we need to reevaluate at one point is how we deal with duplicate entries in maps, which IIRC are allowed in CTAP CBOR

[§ 8. Message Encoding](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#message-encoding) mentions

> All decoders SHOULD reject CBOR that is not validly encoded in the [CTAP2 canonical CBOR encoding form](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#ctap2-canonical-cbor-encoding-form) and SHOULD reject messages with duplicate map keys.

> in others (cbor_decode_attstmt, alg case) we silently overwrite previously observed values

Yes, `alg` is the odd one out there. We could certainly add a demanding `COSE_UNSPEC`.

File: man/fido_cred_set_authdata.3
Comment (LDVG) [reply man/fido_cred_set_authdata.3:164]: I'll add this to my TODOs.

**Review by kongeo — APPROVED** [review]
LGTM!

**Pull request merged**
